### PR TITLE
feat: LFU (Least frequently used) cache

### PR DESCRIPTION
**Pull request by deDSeC00720**
#### Description of Change
LFU Cache implementation
In a fixed size cache, when new item is added and the cache is full
The least frequently used element is evicted from the cache

If there are multiple items with the same frequency of usage,
then the least recently used item is evicted.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

added LFU Cache with tests and usage.

### CODE REVIEW

File: others/lfu_cache.cpp [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+
+/**
+ * @brief main function
+ * @return 0 on exit
+ */
+
+int main() {
+    test();
+    return 0;
+}
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:298]:
```suggestion
}

```

File: others/lfu_cache.cpp [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+void test() {
+    test1();
+    test2();
+}
+
+/**
+ * @brief main function
+ * @return 0 on exit
+ */
+
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:294]:
```suggestion
```

File: others/lfu_cache.cpp [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    assert(cache.get(5) == 50);
+
+    std::cout << "test2 - passed\n";
+}
+
+/**
+ * @brief test method with 2 tests
+ * @return void
+ */
+
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:284]:
```suggestion
```

File: others/lfu_cache.cpp [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+    assert(cache.size() == 0);
+    assert(cache.capacity() == 5);
+    assert(cache.empty());
+    cache.put(1, 10);
+    cache.put(2, 20);
+
+    assert(cache.size() == 2);
+    assert(cache.capacity() == 5);
+    assert(!cache.empty());
+
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:244]: Please add negative numbers since youre using int as the datatype

File: others/lfu_cache.cpp [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+#include <cassert>        // for assert
+#include <iostream>       // for IO operations
+#include <unordered_map>  // for hash map
+
+/**
+ * @namespace
+ * @brief Other algorithms
+ */
+namespace others {
+namespace Cache {
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:31]: This namespace should also be documented

File: others/lfu_cache.cpp [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+
+template <typename T>
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:33]: This class should also be documented

File: others/lfu_cache.cpp [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+    D_Node<T> *next;
+
+    explicit D_Node(T data) : data(data), prev(nullptr), next(nullptr) {}
+};
+
+template <typename K, typename V>
+using CacheNode = D_Node<std::pair<K, V>>;
+
+template <typename K, typename V>
+class LFUCache {
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:47]: This class should be documented as well

File: others/lfu_cache.cpp [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+    std::unordered_map<K, std::pair<CacheNode<K, V> *, int>>
+        node_map;  // maps the key to the node address and frequency
+    std::unordered_map<int, std::pair<CacheNode<K, V> *, CacheNode<K, V> *>>
+        freq_map;  // maps the frequency to doubly linked list
+
+    int minFreq;    // minimum frequency in the cache
+    int _capacity;  // maximum capacity of the cache
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:54]:
```suggestion
    std::unordered_map<K, std::pair<CacheNode<K, V> *, int>>
        node_map;  ///< maps the key to the node address and frequency
    std::unordered_map<int, std::pair<CacheNode<K, V> *, CacheNode<K, V> *>>
        freq_map;  ///< maps the frequency to doubly linked list

    int minFreq;    ///< minimum frequency in the cache
    int _capacity;  ///< maximum capacity of the cache
```

File: others/lfu_cache.cpp [hunk 9]
```diff
@@ -0,0 +1,298 @@
+/**
+ * @file
+ * @brief Implementation for [LFU Cache]
+ * (https://en.wikipedia.org/wiki/Least_frequently_used)
+ *
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:5]:
```suggestion
 *
 * @details
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]
Thank you for your contribution! Let me know if this was for hacktoberfest

**Comment by deDSeC00720** [comment]
> Thank you for your contribution! Let me know if this was for hacktoberfest

Yes, this is for Hacktoberfest. Thank you for reviewing my contribution.

**Comment by realstealthninja** [comment]
> > Thank you for your contribution! Let me know if this was for hacktoberfest
> 
> Yes, this is for Hacktoberfest. Thank you for reviewing my contribution.

youre very welcome. lets get this accepted soon :rocket:

File: others/lfu_cache.cpp [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+ * When the cache is full, the last element in the minimum frequency linked list
+ * is popped.
+ *
+ * @author [Karan Sharma](https://github.com/deDSeC00720)
+ */
+
+#include <cassert>        // for assert
+#include <iostream>       // for IO operations
+#include <unordered_map>  // for hash map
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:25]:
```suggestion
#include <iostream>       // for std::cout
#include <unordered_map>  // for std::unordered_map
```

File: others/lfu_cache.cpp [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+/**
+ * @namespace
+ * @brief Cache algorithm
+ */
+namespace Cache {
+
+/**
+ * @class
+ * @brief Node for a doubly linked list with data, prev and next pointers
+ */
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:42]:
```suggestion
/**
 * @class
 * @brief Node for a doubly linked list with data, prev and next pointers
 * @tparam T type of the data of the node
 */
```

File: others/lfu_cache.cpp [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+template <typename T>
+class D_Node {
+ public:
+    T data;
+    D_Node<T> *prev;
+    D_Node<T> *next;
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:48]:
```suggestion
    T data;           ///< data of the node
    D_Node<T> *prev;  ///< previous node in the doubly linked list
    D_Node<T> *next;  ///< next node in the doubly linked list
```

File: others/lfu_cache.cpp [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+    explicit D_Node(T data) : data(data), prev(nullptr), next(nullptr) {}
+};
+
+template <typename K, typename V>
+using CacheNode = D_Node<std::pair<K, V>>;
+
+/**
+ * @class
+ * @brief LFUCache
+ */
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:59]:
```suggestion
/**
 * @class
 * @brief LFUCache
 * @tparam K type of the key in the LFU
 * @tparam V type of the value in the LFU
 */
```

File: others/lfu_cache.cpp [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+    cache.put(5, 50);
+    cache.put(6, 60);
+
+    assert(cache.get(5) == 50);
+
+    std::cout << "test2 - passed\n";
+}
+
+/**
+ * @brief test method with 2 tests
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:295]:
```suggestion
 * @brief self test implementation
```

File: others/lfu_cache.cpp [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+    test1();
+    test2();
+}
+
+/**
+ * @brief main function
+ * @return 0 on exit
+ */
+int main() {
+    test();
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:308]:
```suggestion
    test(); // runs the self test implementation
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: others/lfu_cache.cpp [hunk 16]
```diff
@@ -0,0 +1,310 @@
...
+        p.first->prev = node;
+        node->next = p.first;
+        p.first = node;
+    }
+
+    /**
+     * @brief increase the frequency of node and push it in the respective list.
+     * @param p_node the node to be updated
+     */
+    void inc_freq(std::pair<CacheNode<K, V> *, int> &p_node) {
```
Comment (aminegh20) [inline others/lfu_cache.cpp:106]: I wouldn't use abbreviations that way. `void increase_frequency` is clearer.

File: others/lfu_cache.cpp [hunk 17]
```diff
@@ -0,0 +1,313 @@
...
+    assert(cache.get(5) == 50);
+
+    std::cout << "test2 - passed\n";
+}
+
+/**
+ * @brief self test implementation
+ * @return void
+ */
+void test() {
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:301]:
```suggestion
static void tests() {
```

File: others/lfu_cache.cpp [hunk 18]
```diff
@@ -0,0 +1,313 @@
...
+    test1();
+    test2();
+}
+
+/**
+ * @brief main function
+ * @return 0 on exit
+ */
+int main() {
+    test();  // run the self test implementation
```
Comment (realstealthninja) [inline others/lfu_cache.cpp:311]:
```suggestion
    tests();  // run the self test implementation
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Comment by deDSeC00720** [comment]
@realstealthninja, what is the reason for making a non member function static?
And should I make test1() and test2() static as well?

**Comment by realstealthninja** [comment]
> @realstealthninja, what is the reason for making a non member function static?
> And should I make test1() and test2() static as well?

Static is used to hide the function from the global namespace. 


Imo there shouldn't be 3 test function just one with comments delimiting the cases

**Comment by deDSeC00720** [comment]
> > @realstealthninja, what is the reason for making a non member function static?
> > And should I make test1() and test2() static as well?
> 
> Static is used to hide the function from the global namespace.
> 
> Imo there shouldn't be 3 test function just one with comments delimiting the cases

Got it.

**Review by realstealthninja — APPROVED** [review]
Lgtm

**Pull request merged**